File: src/spoj/COINS.java
package spoj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;

public class COINS {
    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String s;
        HashMap<Integer, Long> map = new HashMap<>();
        while ((s = in.readLine()) != null) {
            int value = Integer.parseInt(s);
            System.out.println(computeMaxValue(value, map));
        }
    }

    private static long computeMaxValue(int n, HashMap<Integer, Long> map) {
        if (n == 0) return 0;
        if (map.containsKey(n)) return map.get(n);
        long result = Math.max(n, computeMaxValue(n / 2, map) +
                computeMaxValue(n / 3, map) +
                computeMaxValue(n / 4, map));
        map.put(n, result);
        return result;
    }
}

File: cpp/spoj/COINS.cc
#include <math.h>

#include <algorithm>
#include <iostream>
#include <map>
#include <vector>

using namespace std;

map<long long, long long> max_value;

long long get_max(long long n) {
    if (n <= 4) return n;
    if (max_value.find(n) != max_value.end()) {
        return max_value[n];
    }
    long long result = max(n, get_max(n / 2) + get_max(n / 3) + get_max(n / 4));
    max_value[n] = result;
    return result;
}

int main() {
    int n;
    while (cin >> n) {
        cout << get_max(n) << "\n";
    }
    return 0;
}
